File: todo-api-rs/src/framework/rest_api/todo/list.rs
use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};

use super::TodoState;
use crate::adapters::controllers::todo::list::ListController;
use crate::adapters::dtos::todo::list::RunError;
use crate::framework::rest_api::error::ApiError;

pub(super) async fn list_todos(State(state): State<TodoState>) -> impl IntoResponse {
    let controller = ListController::new(state.todo_store);

    let output = match controller.run().await.value() {
        Ok(output) => output,
        Err(err) => {
            let (status, error) = get_error_response_config(err);
            return (status, Json(error)).into_response();
        }
    };

    (StatusCode::OK, Json(output)).into_response()
}

fn get_error_response_config(error: RunError) -> (StatusCode, ApiError<()>) {
    match error {
        RunError::Internal => {
            let error = ApiError::new("LTD-001", "Internal server error");
            (StatusCode::INTERNAL_SERVER_ERROR, error)
        }
    }
}

File: todo-api-js/src/framework/presentation/handlers/todo/list.ts
import { type Context } from 'elysia'

import * as E from '@core/helpers/either'
import { type ParseError } from '@core/helpers/parser'
import { type TodoRepository } from '@application/repositories/todo'
import { ListController, type RunError } from '@adapters/controllers/todo/list'
import { RawInput, type Input, type Output } from '@adapters/dtos/todo/list'
import { type ApiError } from '@framework/presentation/errors'

export type ListHandlerState = { repository: TodoRepository }
export type ListHandlerOutput = Output | ApiError<ParseError<Input>['details']>
export type ListHandlerContext = Context<
  {
    response: Promise<ListHandlerOutput>
  },
  ListHandlerState
>

export async function listHandler(context: ListHandlerContext): Promise<ListHandlerOutput> {
  const input = new RawInput({})
  const controller = new ListController(input, context.store.repository)
  const output = await controller.run()
  if (E.isLeft(output)) {
    const [status, error] = getErrorResponseConfig(output.value)
    context.set.status = status
    return error
  }

  context.set.status = 200
  return output.value
}

function getErrorResponseConfig(error: RunError): [number, ApiError<ParseError<Input>['details']>] {
  switch (error.kind) {
    case 'validation':
      return [400, { code: 'LTD-001', message: 'Invalid input', details: error.details }]
    case 'internal':
      return [500, { code: 'LTD-002', message: 'Internal server error' }]
    default:
      // exhaustive check
      return error
  }
}
